### PR TITLE
Replace black, flake8, pydocstyle, bandit with ruff

### PR DESCRIPTION
**Pull request by jdrew82**
This PR is to update the project linters to replace black, flake8, pydocstyle, and bandit with ruff. This speeds up tests significantly.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -214,17 +247,45 @@ supported_nautobot_versions = [
...
+    "_build",
+    "buck-out",
+    "build",
+    "dist",
+    "nautobot_ssot/integrations/servicenow/third_party"
+]
+
+[tool.ruff.lint]
+select = [
+    "DJ",  # flake8-django
```
Comment (glennmatthews) [inline pyproject.toml:270]: If you're replacing `pydocstyle` then you should enable ruff's `D` rules here as well!

File: nautobot_ssot/integrations/device42/utils/nautobot.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def get_random_color() -> str:
     Returns:
         str: Hex code value for a color with hash stripped.
     """
-    return f"{random.randint(0, 0xFFFFFF):06x}"  # nosec: B311
+    return f"{random.randint(0, 0xFFFFFF):06x}"  # noqa: S311
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply nautobot_ssot/integrations/device42/utils/nautobot.py:33]: Agreed! This is one of those instances were we have code/functionality duplication across integrations that we need to work on refactoring to make more DRY.

File: pyproject.toml [hunk 3]
```diff
@@ -194,12 +167,72 @@ ignore=".venv"
 # No docstrings required for private methods (Pylint default), or for test_ functions, or for inner Meta classes.
 no-docstring-rgx="^(_|test_|Meta$)"
 
-[tool.pylint.messages_control]
-# Line length is enforced by Black, so pylint doesn't need to check it.
-# Pylint and Black disagree about how to format multi-line arrays; Black wins.
-disable = """,
+[tool.pylint.message_control]
+# TODO: re-enable and fix these as time permits
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply pyproject.toml:171]: Yes, most of this is copying from your updates to core. It appears we currently just disable `line-too-long` and `too-few-public-methods` so I can revert this bit back to just those messages.

File: pyproject.toml [hunk 4]
```diff
@@ -214,17 +247,45 @@ supported_nautobot_versions = [
...
+ignore = [
+    "E501",  # pycodestyle: line-too-long
+    "PIE808",  # unnecessary-range-start
+    "RUF012",  # mutable-class-default - see https://github.com/astral-sh/ruff/issues/5243
+    "RUF013",  # prohibits implicit "Optional"
+]
+
+[tool.ruff.lint.isort]
+lines-after-imports = 2
+force-sort-within-sections = true
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply pyproject.toml:288]: Thank you, I was meaning to review that one for updates.

File: pyproject.toml (hunk 1 above)
Comment (jdrew82) [reply pyproject.toml:210]: Good catch, I misread `pycodestyle` on the others and mistakenly thought it was already enabled. Have enabled it now and moved settings from ini into pyproject.toml. For some reason enabling it triggered a bunch of warnings. Not really sure why they didn't come up with our previous implementation of pydocstyle.

File: tasks.py [hunk 5]
```diff
@@ -583,19 +564,28 @@ def pylint(context):
...
-    run_command(context, command)
+@task(
+    help={
+        "action": "One of 'lint', 'format', or 'both'",
+        "autoformat": "Automatically apply formatting recommendations, rather than failing if formatting is incorrect.",
+        "fix": "Automatically apply linting recommendations where possible. May not be able to fix all.",
+        "output_format": "see https://docs.astral.sh/ruff/settings/#output-format",
+    },
+)
+def ruff(context, action="both", autoformat=False, fix=False, output_format="text"):
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply tasks.py]: Got em!

File: nautobot_ssot/integrations/aci/signals.py [hunk 6]
```diff
@@ -42,7 +42,7 @@ def aci_create_tag(apps, **kwargs):
         if ("SITE" in key or "STAGE" in key) and not tag.objects.filter(name=apics[key]).exists():
             tag.objects.update_or_create(
                 name=apics[key],
-                color="".join([random.choice("ABCDEF0123456789") for i in range(6)]),  # nosec
+                color="".join([random.choice("ABCDEF0123456789") for i in range(6)]),  # noqa: S311
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply nautobot_ssot/integrations/aci/signals.py:45]: I like the idea to use ColorChoices so have updated to use that.

File: nautobot_ssot/integrations/device42/utils/nautobot.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdrew82) [reply nautobot_ssot/integrations/device42/utils/nautobot.py]: Updated to use ColorChoices as above.

File: nautobot_ssot/templatetags/render_diff.py [hunk 7]
```diff
@@ -55,12 +55,12 @@ def render_diff_recursive(diff):
                 child_result += render_diff_recursive(child_diffs)
 
             child_result += "</ul></li>"
-        result += format_html("<li>{}<ul>{}</ul></li>", record_type, mark_safe(child_result))  # nosec
+        result += format_html("<li>{}<ul>{}</ul></li>", record_type, mark_safe(child_result))  # noqa: S308
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply nautobot_ssot/templatetags/render_diff.py:57]: Tried this and it just requires moving the ignore up to that line instead. I'm fine doing that if we think it's cleaner but it seems to be the same issue regardless.

**Comment by itdependsnetworks** [comment]
FYI: bandit in ruff is still missing 3 checks, 2 of which are django ones, additionally a bunch have been added recently, so not sure if they have been released.

File: nautobot_ssot/migrations/0006_ssotservicenowconfig.py [hunk 8]
```diff
@@ -41,6 +41,8 @@ def _move_data(apps, schema_editor):
 
 
 class Migration(migrations.Migration):
+    """Migration class."""
```
Comment (itdependsnetworks) [inline nautobot_ssot/migrations/0006_ssotservicenowconfig.py:44]: These should not be needed, we should update the configuration for pydocstyle's ruff equivalant to not need doc strings.

File: nautobot_ssot/integrations/device42/diffsync/adapters/device42.py [hunk 9]
```diff
@@ -677,7 +682,7 @@ def load_ip_addresses(self):
         self.job.logger.info("Loading IP Addresses from Device42.")
         _cfs = self.device42.get_ipaddr_custom_fields()
         for _ip in self.device42.get_ip_addrs():
-            _ipaddr = f"{_ip['ip_address']}/{str(_ip['netmask'])}"
+            _ipaddr = f"{_ip['ip_address']}/{_ip['netmask']}"
```
Comment (itdependsnetworks) [inline nautobot_ssot/integrations/device42/diffsync/adapters/device42.py:685]: How confident are you in this change?

File: nautobot_ssot/integrations/device42/diffsync/adapters/device42.py [hunk 10]
```diff
@@ -702,7 +707,7 @@ def load_ip_addresses(self):
                 _tags = sorted(_ip["tags"].split(",")) if _ip.get("tags") != "" else []
                 new_ip = self.ipaddr(
                     address=_ipaddr,
-                    subnet=f"{_ip['subnet']}/{str(_ip['netmask'])}",
+                    subnet=f"{_ip['subnet']}/{_ip['netmask']}",
```
Comment (itdependsnetworks) [inline nautobot_ssot/integrations/device42/diffsync/adapters/device42.py:710]: Same as previous.

**Comment by glennmatthews** [comment]
> FYI: bandit in ruff is still missing 3 checks, 2 of which are django ones, additionally a bunch have been added recently, so not sure if they have been released.

Any chance you have a link with details about this?

**Comment by itdependsnetworks** [comment]
> > FYI: bandit in ruff is still missing 3 checks, 2 of which are django ones, additionally a bunch have been added recently, so not sure if they have been released.
> 
> Any chance you have a link with details about this?

https://github.com/astral-sh/ruff/issues/1646 and the [faq](https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-flake8) has been very helpful in knowing what is ready vs not as well as the links.

**Comment by glennmatthews** [comment]
Thanks. It's not clear to me how the "missing" `B703` `mark_safe` check is different from the "present" `S308` `mark_safe` check - any insight? I agree that it would be good to get `B610` included.

File: nautobot_ssot/integrations/device42/diffsync/adapters/device42.py (hunk 9 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/device42/diffsync/adapters/device42.py:685]: I'm fairly confident in this. I honestly can't recall why I would cast it to string when using f-strings should automatically do that.

File: nautobot_ssot/integrations/device42/diffsync/adapters/device42.py (hunk 10 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/device42/diffsync/adapters/device42.py:710]: Same response as above.

**Comment by itdependsnetworks** [comment]
I don't know, tbh, here are the relevant links https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b308-mark-safe vs https://bandit.readthedocs.io/en/latest/plugins/b703_django_mark_safe.html

**Comment by jdrew82** [comment]
Closing for now until we figure out how we want to proceed with ruff.

**Pull request closed without merging**